### PR TITLE
feat(iotda): the device resource supports new API and new fields

### PR DESCRIPTION
**Pull request by ruwenqiang123**
**What this PR does / why we need it**:

The device resource supports new API and new fields.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.

```
$ make testacc TEST="./huaweicloud/services/acceptance/iotda" TESTARGS="-run TestAccDevice_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/iotda -v -run TestAccDevice_basic -timeout 360m -parallel 4
=== RUN   TestAccDevice_basic
--- PASS: TestAccDevice_basic (124.67s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/iotda     124.723s
```
```
$ make testacc TEST="./huaweicloud/services/acceptance/iotda" TESTARGS="-run TestAccDevice_derived"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/iotda -v -run TestAccDevice_derived -timeout 360m -parallel 4
=== RUN   TestAccDevice_derived
--- PASS: TestAccDevice_derived (122.90s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/iotda     122.947s
```

* [x] Documentation updated.
* [x] Schema updated.
* [ ] CheckDeleted.

  - **a. During query operation (Read Context)**
    aa. Resource not found
    \>>>>>> Paste the screenshot here <<<<<<

    
  - **b. During delete/disassociate/unbind operation (Delete Context)**
    ba. Resource not found
    \>>>>>> Paste the screenshot here <<<<<<

### CODE REVIEW

**Comment by dengyali0125** [comment]
LGTM

File: docs/resources/iotda_device.md [hunk 1]
```diff
@@ -120,6 +120,31 @@ are allowed: `?'#().,&%@!`.
 
 * `frozen` - (Optional, Bool) Specifies whether to freeze the device. Defaults to **false**.
 
+* `extension_info` - (Optional, Map) Specifies the extended information of the device.
+  The users can be customized the content. The maximum size of the value is `1` KB.
+
+* `shadow` - (Optional, List) Specifies the initial configuration of the device.
+  The [shadow](#device_shadow) structure is documented below.
+
+  You can use this parameter to specify the initial configuration for the device. The system compares the property
```
Comment (jinyangyang222) [inline docs/resources/iotda_device.md:129]:
```suggestion
  -> 1. You can use this parameter to specify the initial configuration for the device. The system compares the property
```

File: docs/resources/iotda_device.md [hunk 2]
```diff
@@ -120,6 +120,31 @@ are allowed: `?'#().,&%@!`.
...
+  values specified by `service_id` and the `desired` section with the default values of the corresponding properties
+  in the product. If they are different, the property values specified by the `shadow` parameter are written to the
+  device shadow.
+
+  -> 1.The value of `service_id` and the properties in the `desired` section must be defined in the product model.
+  <br>2.If you want to config `shadow` data, the `method` value of the properties in the product model must
+  contain **W**.
```
Comment (jinyangyang222) [inline docs/resources/iotda_device.md:136]: Put it together with the previous text

File: docs/resources/iotda_device.md [hunk 3]
```diff
@@ -120,6 +120,31 @@ are allowed: `?'#().,&%@!`.
...
+
+<a name="device_shadow"></a>
+The `shadow` block supports:
+
+* `service_id` - (Required, String) Specifies the service ID of the device.
+  Which is defined in the product model associated with the device.
+
+* `desired` - (Required, Map) Specifies the initial properties data of the device.
+  The each key is a parameter name of a property in the product model.
+  If you want to delete the entire `desired`, please enter an empty Map (**desired = {}**).
```
Comment (jinyangyang222) [inline docs/resources/iotda_device.md:146]:
```suggestion
  If you want to delete the entire `desired`, please enter an empty Map. e.g. **desired = {}**.
```

File: docs/resources/iotda_device.md [hunk 4]
```diff
@@ -157,7 +183,7 @@ resource "huaweicloud_iotda_device" "test" {
   
   lifecycle {
     ignore_changes = [
-      force_disconnect,
+      force_disconnect,extension_info,shadow,
```
Comment (jinyangyang222) [inline docs/resources/iotda_device.md:186]:
```suggestion
      force_disconnect, extension_info, shadow,
```

File: docs/resources/iotda_device.md [hunk 5]
```diff
@@ -146,7 +171,8 @@ $ terraform import huaweicloud_iotda_device.test 10022532f4f94f26b01daa1e424853e
 ```
 
 Note that the imported state may not be identical to your resource definition, due to some attributes missing from the
-API response, security or some other reason. The missing attributes include: `force_disconnect`.
+API response, security or some other reason.
+The missing attributes include: `force_disconnect`,`extension_info`, `shadow`.
```
Comment (jinyangyang222) [inline docs/resources/iotda_device.md:175]:
```suggestion
The missing attributes include: `force_disconnect`, `extension_info`, `shadow`.
```

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 6]
```diff
@@ -114,36 +109,51 @@ func ResourceDevice() *schema.Resource {
...
-
+			"extension_info": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"shadow": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
```
Comment (jinyangyang222) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go:125]: Reason for not using computed attribute added

**Comment by ruwenqiang123** [comment]
```
$ make testacc TEST="./huaweicloud/services/acceptance/iotda" TESTARGS="-run TestAccDevice_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/iotda -v -run TestAccDevice_basic -timeout 360m -parallel 4
=== RUN   TestAccDevice_basic
--- PASS: TestAccDevice_basic (132.72s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/iotda     132.774s
```

File: huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go [hunk 7]
```diff
@@ -454,15 +484,43 @@ func buildDeviceCreateParams(d *schema.ResourceData) *model.AddDeviceRequest {
...
 func buildDeviceUpdateParams(d *schema.ResourceData) *model.UpdateDeviceRequest {
+	extensionInfo := utils.ValueIgnoreEmpty(d.Get("extension_info"))
 	req := model.UpdateDeviceRequest{
 		DeviceId: d.Id(),
 		Body: &model.UpdateDevice{
-			DeviceName:  utils.StringIgnoreEmpty(d.Get("name").(string)),
-			Description: utils.StringIgnoreEmpty(d.Get("description").(string)),
+			DeviceName:    utils.StringIgnoreEmpty(d.Get("name").(string)),
+			Description:   utils.StringIgnoreEmpty(d.Get("description").(string)),
+			ExtensionInfo: &extensionInfo,
```
Comment (deer-hang) [inline huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go]: The return value of method `utils.ValueIgnoreEmpty()` may be nil, and this line of code has the risk of null pointer.

**Comment by ruwenqiang123** [comment]
```
$ make testacc TEST="./huaweicloud/services/acceptance/iotda" TESTARGS="-run TestAccDevice_derived"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/iotda -v -run TestAccDevice_derived -timeout 360m -parallel 4
=== RUN   TestAccDevice_derived
--- PASS: TestAccDevice_derived (126.58s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/iotda     126.641s
```

**Pull request merged**